File: src/sources/fb-messenger.go
package sources

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"

	"github.com/nt92/memex/src/lib"
	"github.com/nt92/memex/src/schema"
)

type MessengerMessage struct {
	Sender   string `json:"sender_name"`
	Time     int    `json:"timestamp_ms"`
	Content  string `json:"content"`
	Type     string `json:"type"`
	IsUnsent bool   `json:"is_unsent"`
}

type MessengerMessages struct {
	List []MessengerMessage `json:"messages"`
}

const messengerPath = "./../data/fb-messenger/messages/messages/inbox/"
const messengerPrefix = "msgr"

func getMessengerRecords() schema.RecordInfo {
	var messengerFiles = getMessengerFileList()
	messageRecordMap := make(schema.RecordInfo)

	// for threadIndex, file := range messengerFiles {
	for threadIndex := 0; threadIndex <= 100; threadIndex++ {
		// currentJsonMessage, _ := os.Open(file)
		currentJsonMessage, _ := os.Open(messengerFiles[threadIndex])
		defer currentJsonMessage.Close()

		byteValue, _ := ioutil.ReadAll(currentJsonMessage)

		var messageList MessengerMessages
		err := json.Unmarshal([]byte(byteValue), &messageList)
		if err != nil {
			log.Fatal(err)
		}

		for messageIndex, message := range messageList.List {
			messageID := fmt.Sprintf("%s-%d-%d", messengerPrefix, threadIndex, messageIndex)
			tokens := lib.GetTokenFrequencyMap(message.Content)

			messageRecord :=
				schema.Record{
					ID:             messageID,
					Title:          "Messenger Message from " + message.Sender,
					Content:        message.Sender + ": " + message.Content,
					Time:           message.Time / 1000,
					TokenFrequency: tokens,
				}
			messageRecordMap[messageID] = messageRecord
		}
	}

	return messageRecordMap
}

func getMessengerTokens() schema.TokenIndex {
	var messengerFiles = getMessengerFileList()
	tokenMap := make(schema.TokenIndex)

	// for threadIndex, file := range messengerFiles {
	for threadIndex := 0; threadIndex <= 100; threadIndex++ {
		// currentJsonMessage, _ := os.Open(file)
		currentJsonMessage, _ := os.Open(messengerFiles[threadIndex])
		defer currentJsonMessage.Close()

		byteValue, _ := ioutil.ReadAll(currentJsonMessage)

		var messageList MessengerMessages
		err := json.Unmarshal([]byte(byteValue), &messageList)
		if err != nil {
			log.Fatal(err)
		}

		for messageIndex, message := range messageList.List {
			messageID := fmt.Sprintf("%s-%d-%d", messengerPrefix, threadIndex, messageIndex)
			tokens := lib.GetTokenFrequencyMap(message.Content)

			for token := range tokens {
				_, ok := tokenMap[token]
				if ok {
					tokenMap[token] = append(tokenMap[token], messageID)
				} else {
					tokenMap[token] = []string{messageID}
				}
			}
		}
	}

	return tokenMap
}

func getMessengerFileList() []string {
	var files []string
	err :=
		filepath.Walk(
			messengerPath,
			func(path string, info os.FileInfo, err error) error {
				if filepath.Ext(path) != ".json" {
					return nil
				}
				files = append(files, path)
				return nil
			})
	if err != nil {
		panic(err)
	}

	return files
}

// TODO: possibly remove records for "You are now connected on Messenger"
// TODO: figure out how to handle & index names properly
// TODO: handle emojis somehow

File: src/sources/fb_messenger.py
import glob
import json

from src.lib.database import Database

import re

messenger_path = "./data/fb-messenger/messages/messages/inbox/*/*.json"
# messenger_path = "./data/fb-messenger/messages/messages/inbox/mackenziepatel_8sbrouv7sg/*.json"
messenger_prefix = "msgr"


def get_messenger_records_and_tokens(db: Database):
    messenger_json_list = glob.glob(messenger_path)

    for file_index, file in enumerate(messenger_json_list, len(messenger_json_list) - 1):
        with open(file, "r") as f:
            data = json.load(f)

            messages_with_content = (messages for messages in data["messages"] if 'content' in messages)
            for message_index, message in enumerate(messages_with_content):
                # msgr saves special char content in a weird way so we have to decode it
                message_content = re.sub(r'[\xc2-\xf4][\x80-\xbf]+',
                                         lambda m: m.group(0).encode('latin1').decode('utf8'), message["content"])

                source = messenger_prefix
                title = "Messenger Thread with " + data["title"]
                content = message["sender_name"] + ": " + message_content
                time = message["timestamp_ms"]

                db.save_record(source, title, content, time, "")
